(TK-192) Support Arrays

fpringvaldsen opened the pull request:
This PR adds support for Arrays, Hashes, and rendered text (the text of the value as it should appear in the configuration file) for the value parameter.

This PR is based on https://github.com/puppetlabs/puppetlabs-hocon/pull/7, so that should get merged before this is reviewed.

This PR will also require the changes in https://github.com/puppetlabs/ruby-hocon/pull/55 being released and the gemfile updated before it can be merged.

fpringvaldsen commented:
This should cover the last bit of work for [TK-167](https://tickets.puppetlabs.com/browse/TK-167).

There's still some more work that needs to be done before this module's next release. We should get rid of nodesets that run with Ruby 1.8, and ensure we feel both the spec and acceptance testing coverage is sufficient.

fpringvaldsen commented:
This should be ready for review now (I've rebased and updated the ruby-hocon gem dependency to 0.9.0).

cprice404 commented on README.markdown:
```diff
@@ -77,6 +77,19 @@ hocon_setting {'sample nested setting':
+    In this case, `value` must be a string with no leading or trailing whitespace, newlines, or comments that contains a valid HOCON value, and the
+    `type` parameter must be set to `'text'`. This is an advanced use case, and will not be necessary for most users.
+    
+    Aside from these two cases, the `type` parameter does not need to be set.
```
My understanding was that 'type' could be set to 'map', 'array', etc., even though it isn't strictly required.  If that's the case, we should list out the legal values for the enumeration as the first bit of the docs here, just to avoid confusion.

cprice404 replied:
Could we also add an example or two the examples you show earlier in the README?  e.g. one example of using a map value, one of using the 'raw text' feature, and then... maybe one showing the edge case thing for a single-item array?  :crying_cat_face:

cprice404 commented on spec/unit/puppet/provider/conf_setting/ruby_spec.rb:
```diff
@@ -212,14 +212,106 @@ def validate_file(expected_content,tmpfile = tmpfile)
+      provider = described_class.new(resource)
+      provider.create
+      expect(provider.exists?).to be true
+      expect(provider.value).to eql({"a" => 1, "b" => 2})
```
Just to validate your logic here:

This test only tests that the value of the Puppet resource ends up being set to a map, right?  It doesn't actually test that the module can successfully pass a map into the HOCON API, and validate that we can read the same map back out of the HOCON API?  Do you have coverage for that somewhere else in the module?

I'm thinking you probably don't need to have a test for roundtripping a map value to a file, because we have tests for that in the gem, but it might still be worth just testing that the Puppet module is interacting with the HOCON gem API successfully.

fpringvaldsen replied:
This test should cover that use case, as the `create` method handles adding the set value into the file and the `value` method works by using the HOCON API to grab the value of the desired setting from the config file.

cprice404 replied:
Ah, ok.  great.

cprice404 commented:
+1 other than my two suggestions / questions.

igalic commented on lib/puppet/type/hocon_setting.rb:
```diff
@@ -22,8 +22,36 @@
+          if @shouldorig.is_a?(Array) and @shouldorig.size > 1
+            raise "Array provided, but type specified as #{@resource[:type]}"
+          end
+      end
```
i'm finding this validation here really, really confusing.

cprice404 replied:
Maybe would help to add a few comments to the code?

igalic replied:
i was reading it a couple of times looking for a way to simplify it, or collapse the paths, but there's too much going on here in order to be able to do that :\

fpringvaldsen replied:
Yeah, I agree it's somewhat confusing.

This method handles setting the `type` parameter, but in the provider code, the only specific cases that matter are when it is set to `'array'` or `'text'`. I will probably remove the logic that actually sets that parameter here since it's really not necessary, and just do some basic type validation here with the setting if it happens to be present, since that's what I think should happen anyway.

fpringvaldsen commented:
@cprice404 I've addressed your comments!

I've also completely redone the validation of the `type` parameter. Essentially, if the `type` parameter exists, the validation function now validates that the value is of the correct type and throws an exception if it isn't. It also throws an exception if the `type` parameter is an invalid string. Hopefully this code is a lot more straightforward now and addresses @igalic's feedback.

cprice404 commented:
LGTM, merging in the interest of getting this in before you leave for vacation.  If @igalic or anyone else has additional feedback we can add another PR.
